lib/reporting/reports/customers/base.rb

```diff
@@ -5,24 +5,57 @@ module Reports
     module Customers
       class Base < ReportTemplate
         def query_result
-          filter Spree::Order.managed_by(@user)
+          filter(Spree::Order.managed_by(@user)
             .distributed_by_user(@user)
             .complete.not_state(:canceled)
+            .group_by do |order|
+              {
+                first_name: order.billing_address.firstname,
+                last_name: order.billing_address.lastname,
+                billing_address: order.billing_address.address_and_city,
+                email: order.email,
+                phone: order.billing_address.phone,
+                hub_id: order.distributor_id,
+                shipping_method_id: order.shipping_method&.id,
+              }
+            end.values)
         end
 
+        # rubocop:disable Metrics/AbcSize
+        def columns
+          {
+            first_name: proc { |orders| orders.first.billing_address.firstname },
+            last_name: proc { |orders| orders.first.billing_address.lastname },
+            billing_address: proc { |orders| orders.first.billing_address.address_and_city },
+            email: proc { |orders| orders.first.email },
+            phone: proc { |orders| orders.first.billing_address.phone },
+            hub: proc { |orders| orders.first.distributor&.name },
+            hub_address: proc { |orders| orders.first.distributor&.address&.address_and_city },
+            shipping_method: proc { |orders| orders.first.shipping_method&.name },
+            total_orders: proc { |orders| orders.count },
+            total_incl_tax: proc { |orders| orders.sum(&:total) },
+            last_completed_order_date: proc { |orders| last_completed_order_date(orders) },
+          }
+        end
+        # rubocop:enable Metrics/AbcSize
+
         def filter(orders)
-          filter_to_supplier filter_to_distributor filter_to_order_cycle orders
-        end
-
-        def filter_to_supplier(orders)
-          if params[:supplier_id].to_i > 0
-            orders.select do |order|
-              order.line_items.includes(:product)
-                .where("spree_products.supplier_id = ?", params[:supplier_id].to_i)
-                .references(:product)
-                .count
-                .positive?
-            end
+          filter_to_completed_at filter_to_distributor filter_to_order_cycle orders
+        end
+
+        def skip_duplicate_rows?
+          true
+        end
+
+        private
+
+        def filter_to_completed_at(orders)
+          if params[:q] &&
+             params[:q][:completed_at_gt].present? &&
+             params[:q][:completed_at_lt].present?
+            orders.where("completed_at >= ? AND completed_at <= ?",
+                         params[:q][:completed_at_gt],
+                         params[:q][:completed_at_lt])
           else
             orders
           end
```

Comment:
Here are some ideas:

* `fetch` is easier to reach into an array.
* Variable names add to understanding the code.
* A guard statement avoids indentation.
* A guard statement puts the focus on the main logic of the method.
* Using a value range is a bit shorter and clearer while doing the same.

```suggestion
          min = params.fetch(:q, :completed_at_gt)
          max = params.fetch(:q, :completed_at_lt)

          return orders if min.blank? || max.blank?

          orders.where(completed_at: [min..max])
```

Reply:
oh wooo. Much more readable. Thanks for that! 🙏